src/Stratis.Bitcoin/P2P/Peer/NetworkPeerServer.cs

```diff
@@ -32,8 +32,14 @@ public class NetworkPeerServer : IDisposable
         /// <summary>The parameters that will be cloned and applied for each peer connecting to <see cref="NetworkPeerServer"/>.</summary>
         public NetworkPeerConnectionParameters InboundNetworkPeerConnectionParameters { get; set; }
 
-        /// <summary>Maximal number of inbound connection that the server is willing to handle simultaneously.</summary>
-        public int MaxConnections { get; set; }
+        /// <summary>Maximum number of inbound connection that the server is willing to handle simultaneously.</summary>
+        public const int MaxConnectionThreshold = 125;
+
+        /// <summary>
+        /// The amount of time in seconds we should stall before checking if we can accept incoming connections again.
+        /// This happens once we have reached the maximum connection threshold.
+        /// </summary>
+        public const int MaxConnectionThresholdStallTime = 60;
```

Comment:
This feels conceptually wrong, it basically introduces a denial of service vector. As it seems that.net implementation does not allow us to reject the connection, we should accept it, check if we are okay with that, which in this case means that the number of existing connections is reasonable, but it could also mean that the incoming peer is not banned

Reply:
@Aprogiena are you saying that we **should** connect first, then check the connection count, **then** if it is more than the maximum, disconnect?

Reply:
It sounds strange, but yes. It's strange because usually you would want to keep your invariants, in this is the maximum number of connections being the maximum, but sadly this is hard to do in.net as we don't have that reject. So we'll allow one more connection over that maximum.

Reply:
OK so in this case, we will connect, create the NetworkPeer then check the limit, then disconnect? This means that we will then always connect and reconnect when the limit has been reached...

```
while (!this.serverCancel.IsCancellationRequested)
{
    // Used to record any errors occurring in the thread pool task.
    Exception error = null;

    TcpClient tcpClient = await Task.Run(() =>
    {
        try
        {
            Task<TcpClient> acceptTask = this.tcpListener.AcceptTcpClientAsync();
            acceptTask.Wait(this.serverCancel.Token);
            return acceptTask.Result;
        }
        catch (Exception exception)
        {
            // Record the error.
            error = exception;
            return null;
        }
    }).ConfigureAwait(false);

    // Raise the error.
    if (error != null)
        throw error;

    this.logger.LogTrace("Connection accepted from client '{0}'.", tcpClient.Client.RemoteEndPoint);

    NetworkPeer networkPeer = this.networkPeerFactory.CreateNetworkPeer(this.Network, tcpClient, this.CreateNetworkPeerConnectionParameters());

    if (this.ConnectedNetworkPeers.Count > MaxConnectionThreshold)
    {
        this.logger.LogTrace("Maximum connection threshold [{0}] reached, stalling accepting of incoming connections.", MaxConnectionThreshold);
        networkPeer.Disconnect("Maximum connection threshold [{0}] reached, stalling accepting of incoming connections.", MaxConnectionThreshold);

        await Task.Delay(MaxConnectionThresholdStallTime, this.serverCancel.Token).ConfigureAwait(false);
        continue;
    }

    this.ConnectedNetworkPeers.Add(networkPeer);
    networkPeer.StateChanged.Register(this.OnStateChangedAsync);

    this.AddClientConnection(networkPeer.Connection);
}
```

Reply:
We may not need to create network peer, we can disconnect tcp client.

Reply:
Ah and we can send a reject payload message if thats what you mean by `reject`

Reply:
No I didn't mean reject message from bitcoin, I meant reject socket function

Reply:
Asked me on Skype about the denial of service

Reply:
@Aprogiena see latest commit... Is this what you meant?

Reply:
almost, except that we need to remove this
```
await Task.Delay(MaxConnectionThresholdStallTime, this.serverCancel.Token).ConfigureAwait(false);
```

and we should then modify the comment accordingly